Add a user guide with compiler choices (#2366)
lfortran/lfortran: doc/src/usage.md

# LFortran User Guide

## About LFortran

> LFortran is an implementation of the Fortran 2018 standard with
> some added extensions.  It works on Linux, most BSDs and on Windows.

Reviewer: ```suggestion
LFortran's goal is an implementation of the latest Fortran standard (currently F2018) with
some added extensions.  It works on Linux, macOS, most BSDs and on Windows.
```

Author: LFortran doesn't have a goal -- we do :)  It's not a big concern.

Reviewer: Go ahead and rewrite it. I am just trying to say that today LFortran is not an implementation of F2018, but we want to be, that's our goal.

The LFortran compiler consists of the following components:

* The LFortran compiler itself.
* A runtime library supporting language elements and intrinsic functions.
* Several module definitions which define the external environment and
  supply particular details about the computer running the program.
* The capability to read additional module definitions which define
  the interfaces to external libraries such as OpenMPI.
* Generation of executable code with the LLVM compiler infrastructure.

> * Generation of C++, WebAssembly or Julia code.

Reviewer: ```suggestion
* Generation of C, C++, WebAssembly or Julia code.
```

> * Source file formatting (`lfortran fmt`).

Reviewer: ```suggestion
* Interactive usage via LLVM JIT
* Source file formatting (`lfortran fmt`).
```

## Standards

> The LFortran compiler is currently (mid-2023) incomplete but is intended

Reviewer: ```suggestion
The LFortran compiler is currently (mid-2023) incomplete (alpha stage) but is intended
```

to support ISO Fortran 2018, ISO/IEC 1539:2018.

## Extensions

The extensions are currently in development and are planned to include:

* Global Scope (statements outside of a program block).

> * Interactive Fortran (in beta).

Reviewer: ```suggestion
* Interactive Fortran (in alpha).
```

* Jupyter integration.

> * Support for FPUs.

Reviewer: What is an FPU? I don't know if you meant this:
```suggestion
* Support for GPUs and other accelerators
```

* --openmp, Enable OpenMP

### Compiler text outputs

* -S, Emit assembly, do not assemble or link
* -v, Be more verbose
* --time-report, Show compilation time report
* --no-warnings, Turn off all warnings
* --no-error-banner, Turn off error banner
* --error-format <value>, Control how errors are produced (human, short)

### Compiler binary outputs

* -c, Compile and assemble, do not link
* -o <value>, Specify the file to place the output into
* -J <value>, Where to save mod files
* -S, Emit assembly, do not assemble or link
* --static, Create a static executable
* --generate-object-code, Generate object code into .o files

### Compiler debugging

A number of command-line options select various text outputs useful
for debugging the compiler.  See `lfortran --help` for a list.

## Examples

The following commands and code demonstrate basic operation of the compiler.

```
lfortran helloworld.f90
Hello World!

lfortran -o hw helloworld.f90
./hw
Hello World!

cat helloworld.f90
program hello_world
    implicit none
    write (*, *) 'Hello World!'
end program hello_world

```

Here is a simple example with a module:

```
lfortran -c varray.f90

> lfortran usev.f90 varray.o

Reviewer: ```suggestion
lfortran usev.f90
```

>   sum is    72.0000000

Reviewer: ```suggestion
 sum is  7.20000000e+01
```

Author: This still needs a space in front of "sum", as Fortran, as you know, has the ancient capability of "carriage control" using the first character of a line as an indicator to a line printer about form feeding or printing the next line.

This also matches my test results.

cat varray.f90

> module varray
>     integer :: nsize
>     real, allocatable, dimension(:) :: A
> end module varray

Reviewer: ```suggestion
module varray
    integer :: nsize
end module varray
```

cat usev.f90

> program usev
>     use varray
>     integer :: i
> 
>     nsize = 8
>     allocate(A(nsize))
> 
>     do i = 1, nsize
>         A(i) = 2.0*i
>     end do
> 
>     print *, " sum is ", (A(1)+A(nsize))*nsize/2.0
>     deallocate(A)
> end program usev

Reviewer: ```suggestion
program usev
    use varray
    real, allocatable, dimension(:) :: A
    integer :: i
    nsize = 8
    allocate(A(nsize))
    do i = 1, nsize
        A(i) = 2.0*i
    end do
    print *, " sum is ", (A(1)+A(nsize))*nsize/2.0
    deallocate(A)
end program usev
```

```

> (Note that as of lfortran version 0.19.0 this does not yet work.)

Reviewer: After the above changes it works.
```suggestion
```

The compile command for the module requires `-c` to avoid automatic running
of the code.

> The compile command for the program requires to add the `.o` object file
> for linking.

Reviewer: Actually, by default it doesn't, since the compilation happens at the very end, the mod file contain all the ASR code of the module, so no need for ".o". We do support separate compilation via the `--generate-object-code` option.
```suggestion
```
